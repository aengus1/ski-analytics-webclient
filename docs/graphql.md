# GraphQL
##1. Overview
GraphQL encourages a very different approach to state management than the 
Redux / NGRX approach.  It is considerably simpler - suggesting that queries and components
are co-located so that the code is easier to follow.  No need for services, actions, reducers,
stores etc etc.  

We are using a mixture of GraphQL and NGRX styles in this application.  I know that this is
annoying and adds a bunch of complexity.  The reasons for this are largely historical but 
can be summarized as:
- graphql doesn't support binary payloads.  Therefore the activity module with it's protobuf
implementation is not a good candidate for using graphql.
- what graphql is great at is providing a simple and flexible interface for searching and for
updating data.  

###In this application we are using graphql for:
- retrieving / updating user settings
- retrieving / updating activity editable fields (planned)
- activity search (in progress)


### What are the problems with GraphQL and particularly with mixing GraphQL with NGRX?

- How do you share data between components?  In NgRx this is straightforward.  Everything
is persisted to the store and can be accessed from there.  
     - We could do this with data retrieved through GraphQL too.. and just use GraphQL for
     querying.
- Developer needs to spend time figuring out what pattern is used before diving in


### Code Gen
To generate model classes from the graphql schema:  `npm run generate`  This outputs to the 
/generated/graphql.ts class
