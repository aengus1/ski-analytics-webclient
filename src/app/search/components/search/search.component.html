<app-search-bar></app-search-bar>
<app-search-results [searchResults]="searchResults"></app-search-results>
