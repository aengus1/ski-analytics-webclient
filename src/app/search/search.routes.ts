import {NgModule} from '@angular/core';
import {RouterModule, Routes} from '@angular/router';
import {AuthGuard} from '../auth/guards/auth.guard';
import {SearchRootComponent} from './components/search-root/search-root.component';
import {ViewSearchPageComponent} from './components/view-search-page/view-search-page.component';

const searchRoutes: Routes = [
  {
    path: 'search', component: SearchRootComponent, children: [
      {path: ':params',
        component: ViewSearchPageComponent,
        canActivate: [AuthGuard]
      },
      {path: '',
        component: ViewSearchPageComponent,
        canActivate: [AuthGuard]
      }
    ]
  }
];
@NgModule({
  imports: [
    RouterModule.forChild(searchRoutes)
  ],
  exports: [
    RouterModule
  ]
})
export class SearchRoutesModule {}
