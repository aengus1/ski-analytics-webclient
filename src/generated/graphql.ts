
export type Maybe<T> = T | null;
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string,
  String: string,
  Boolean: boolean,
  Int: number,
  Float: number,
};

export type Mutation = {
  addUser?: Maybe<User>,
  saveHeight?: Maybe<User>,
  saveWeight?: Maybe<User>,
  saveGender?: Maybe<User>,
  saveHrZones?: Maybe<User>,
};


export type MutationAddUserArgs = {
  userId: Scalars['ID'],
  height?: Maybe<Scalars['Int']>,
  weight?: Maybe<Scalars['Int']>,
  gender?: Maybe<Scalars['Int']>,
  hrZones?: Maybe<Array<Maybe<Scalars['Int']>>>
};


export type MutationSaveHeightArgs = {
  userId: Scalars['ID'],
  height: Scalars['Int']
};


export type MutationSaveWeightArgs = {
  userId: Scalars['ID'],
  weight: Scalars['Int']
};


export type MutationSaveGenderArgs = {
  userId: Scalars['ID'],
  gender: Scalars['String']
};


export type MutationSaveHrZonesArgs = {
  userId: Scalars['ID'],
  hrZones: Array<Maybe<Scalars['Int']>>
};

export type Query = {
  getUser: User,
};


export type QueryGetUserArgs = {
  id?: Maybe<Scalars['Int']>
};

export type Schema = {
  query?: Maybe<Query>,
  mutation?: Maybe<Mutation>,
};

export type User = {
  userId: Scalars['ID'],
  gender?: Maybe<Scalars['String']>,
  height?: Maybe<Scalars['Int']>,
  weight?: Maybe<Scalars['Int']>,
  hrZones?: Maybe<Array<Maybe<Scalars['Int']>>>,
};
