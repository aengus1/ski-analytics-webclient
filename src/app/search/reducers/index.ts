import {ActionReducerMap, createFeatureSelector, createSelector} from '@ngrx/store';

import * as fromSearch from './search.reducer';
import * as fromRoot from '../../reducers';
import * as fromLayout from '../../shared/layout/reducers/layout.reducer';



export interface SearchState {
  search: fromSearch.State;
  error: any;
}
export const reducers: ActionReducerMap<SearchState> = {
  search: fromSearch.reducer,
  error: null
};


export interface State extends fromRoot.State {
  search: SearchState;
}


/**
 * The createFeatureSelector function selects a piece of state from the root of the state object.
 * This is used for selecting feature states that are loaded eagerly or lazily.
 */
export const getSearchState = createFeatureSelector<SearchState>('search');

export const getLayoutState = createFeatureSelector<fromLayout.State>('layout');

export const getSearchEntitiesState = createSelector(
  getSearchState,
  state => state.search
);

// export const selectSearchResults =  (state: SearchState) => state.search.searchResults;

export const selectSearchResults = createSelector(
  getSearchEntitiesState,
  state => state.searchResults
);




