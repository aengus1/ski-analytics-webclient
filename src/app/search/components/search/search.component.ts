import {Component, EventEmitter, Input, OnInit, Output} from '@angular/core';
import {ActivitySearchResult} from '../../../../generated/graphql';
import {MessageEvent} from '../../../shared/utils';
import {ActivitySidebarType} from '../../../activity/actions/activity.actions';

@Component({
  selector: 'app-search-module',
  templateUrl: './search.component.html',
  styleUrls: ['./search.component.scss']
})
export class SearchComponent implements OnInit {

  @Input()
  public searchResults: ActivitySearchResult[];

  @Output()
  public messageEvent = new EventEmitter<MessageEvent<string | ActivitySidebarType>>();

  constructor() { }

  ngOnInit(): void {
  }

}
