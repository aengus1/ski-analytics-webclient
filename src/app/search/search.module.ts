import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {SharedModule} from '../shared/shared.module';
import {ChartModule} from '../chart/chart.module';
import {environment} from '../../environments/environment';
import {StoreDevtoolsModule} from '@ngrx/store-devtools';
import {DefaultRouterStateSerializer, StoreRouterConnectingModule} from '@ngrx/router-store';
import {NgbModule} from '@ng-bootstrap/ng-bootstrap';
import {FormsModule} from '@angular/forms';
import {LoggerService} from '../shared/services/logger.service';
import { SearchRootComponent } from './containers/search-root/search-root.component';
import {SearchRoutesModule} from './search.routes';
import {SearchService} from './services/search.service';
import { SearchBarComponent } from './components/search-bar/search-bar.component';
import { SearchResultsComponent } from './components/search-results/search-results.component';
import { SearchComponent } from './components/search/search.component';
import { ViewSearchResultComponent } from './containers/view-search-result/view-search-result.component';
import { SearchContainerComponent } from './containers/search-container/search-container.component';
import {EffectsModule} from '@ngrx/effects';
import {SearchEffects} from './effects/search.effects';
import {StoreModule} from '@ngrx/store';
import {reducers} from '../search/reducers';
import { PageNumberDisplayComponent } from './components/page-number-display/page-number-display.component';

@NgModule({
  imports: [
    CommonModule,
    SharedModule,
    ChartModule,
    SearchRoutesModule,
    !environment.production ? StoreDevtoolsModule.instrument() : [],
    StoreModule.forFeature('search', reducers),
    StoreRouterConnectingModule.forRoot({ serializer: DefaultRouterStateSerializer,
      stateKey: 'router' // name of reducer key
    }),
    EffectsModule.forFeature([SearchEffects]),
    NgbModule,
    FormsModule
  ],
  declarations: [
    SearchRootComponent,
    SearchComponent,
    SearchBarComponent,
    SearchResultsComponent,
    SearchComponent,
    ViewSearchResultComponent,
    SearchContainerComponent,
    PageNumberDisplayComponent
  ],
  providers: [
    SearchService, LoggerService
  ],
  exports: [
    SearchBarComponent,
    SearchResultsComponent,
    SearchComponent,
    SearchRootComponent,
    ViewSearchResultComponent,
    SearchContainerComponent,
    PageNumberDisplayComponent
  ]
})
export class SearchModule { }
