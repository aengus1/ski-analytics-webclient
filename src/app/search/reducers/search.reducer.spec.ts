import * as fromSearch from './search.reducer';
import {reducer} from './search.reducer';
import {ActivitySearchResult, Criteria, DataType, Maybe, Operator, OrderInfo, PageInfo, Scalars} from '../../../generated/graphql';

describe('Search Reducer', () => {

  const searchResults: ActivitySearchResult[] = new Array<ActivitySearchResult>();

  const initState: fromSearch.State = {
    searchResults: new Array<ActivitySearchResult>(),
    ids: [],
    entities: null
  };

  it('should return the default state', () => {
    const action = {} as any;
    const result = reducer(undefined, action);
    (<any>expect(result)).toMatchSnapshot();
  });

  // it('should call search service when search request is made', () => {
  //   const criteria: Criteria[] = [
  //     {
  //       name: 'test',
  //       value: 'MYTEST',
  //       operator: Operator.Eq,
  //       type: DataType.String
  //     }
  //   ];
  //
  //   const pageInfo: PageInfo = {
  //     pageSize: 40,
  //     pageNumber: 1
  //   };
  //   const order: OrderInfo = {
  //     attribute: 'test',
  //     asc: true
  //   };
  //   const action = new SearchRequest(criteria, pageInfo, order);
  //   const result = reducer(initState, action);
  //   expect(result.searchResults.length
  //   expect(Object.keys(result.entities).length).toEqual(1, 'entities dictionary should not be empty');
  //   expect(result.ids.length).toEqual(1, 'id array should not be empty');
  //   console.log(result.entities[result.selectedActivityId]);
  //   expect(result).toEqual(activity, 'unfiltered activity is not the same as the loaded activity ');
  //
  // });



});


