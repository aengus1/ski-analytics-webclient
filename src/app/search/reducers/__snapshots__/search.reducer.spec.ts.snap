// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Reducer should return the default state 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
  "searchResults": Array [],
}
`;
